crates/server/src/server/h3_handler.rs

```diff
@@ -0,0 +1,144 @@
+// Copyright 2015-2021 Benjamin Fry <benjaminfry@me.com>
+//
+// Licensed under the Apache License, Version 2.0, <LICENSE-APACHE or
+// http://apache.org/licenses/LICENSE-2.0> or the MIT license <LICENSE-MIT or
+// http://opensource.org/licenses/MIT>, at your option. This file may not be
+// copied, modified, or distributed except according to those terms.
+
+use std::{io, net::SocketAddr, sync::Arc};
+
+use bytes::{Buf, Bytes};
+use drain::Watch;
+use futures_util::lock::Mutex;
+use h3::server::RequestStream;
+use h3_quinn::BidiStream;
+use tracing::{debug, warn};
+use trust_dns_proto::{
+    error::ProtoError, h3::h3_server::H3Connection, http::Version, https::HttpsError, rr::Record,
+};
+
+use crate::{
+    authority::MessageResponse,
+    server::{
+        request_handler::RequestHandler, response_handler::ResponseHandler, server_future,
+        Protocol, ResponseInfo,
+    },
+};
+
+pub(crate) async fn h3_handler<T>(
+    handler: Arc<T>,
+    mut connection: H3Connection,
+    src_addr: SocketAddr,
+    _dns_hostname: Option<Arc<str>>,
+    shutdown: Watch,
+) -> Result<(), ProtoError>
+where
+    T: RequestHandler,
+{
+    // TODO: we should make this configurable
+    let mut max_requests = 100u32;
+
+    // Accept all inbound requests sent over the connection.
+    loop {
+        let (_, mut stream) = tokio::select! {
+            result = connection.accept() => match result {
+                Some(Ok(next_request)) => next_request,
+                Some(Err(err)) => {
+                    warn!("error accepting request {}: {}", src_addr, err);
+                    return Err(err);
+                }
+                None => {
+                    break;
+                }
+            },
+            _ = shutdown.clone().signaled() => {
+                // A graceful shutdown was initiated.
+                break;
+            },
+        };
+
+        let request = match stream
+            .recv_data()
+            .await
+            .map_err(|e| ProtoError::from(format!("h3 stream receive data failed: {e}")))?
+        {
+            Some(mut request) => request.copy_to_bytes(request.remaining()),
```

Comment:
This is a bit unfortunate, but our decoder only takes a `&[u8]`.

Reply:
as apposed to taking an `&Bytes` or similar? That's worth a discussion. There is this unresolved issue that we could consider solutions in: #1612

Reply:
This would probably need `impl Buf` or `impl Read`, I will note it in #1612.